timemachine/fe/single_topology.py

```diff
@@ -1393,6 +1130,184 @@ def _align_nonbonded_pair_list(self):
             beta=src_beta,
         )
 
+    def _process_ixn_group_idxs_and_params(
+        self, src_row_or_col_idxs, src_params, dst_row_or_col_idxs, dst_params, a_to_c, b_to_c, cutoff
+    ):
+        # note a_to_c, b_to_c are *combined* with the atoms
+        combined_kv = dict()  # key: atom_idxs indexed in th combined, val: params
+        for idx in src_row_or_col_idxs:
+            q, s, e, w = src_params[idx]
+            new_idx = a_to_c[idx]
+            combined_kv[new_idx] = [(q, s, e, w), (q, s, e, cutoff)]
+        for idx in dst_row_or_col_idxs:
+            q, s, e, w = dst_params[idx]
+            new_idx = b_to_c[idx]
+            if new_idx in combined_kv:
+                combined_kv[new_idx][1] = (q, s, e, w)
+            else:
+                combined_kv[new_idx] = [(q, s, e, cutoff), (q, s, e, w)]
+
+        all_idxs = []
+        src_params = []
+        dst_params = []
+
+        for idx, (src_p, dst_p) in combined_kv.items():
+            all_idxs.append(idx)
+            src_params.append(src_p)
+            dst_params.append(dst_p)
+
+        return (
+            np.array(all_idxs, dtype=np.int32),
+            jnp.array(src_params, dtype=np.float64),
+            jnp.array(dst_params, dtype=np.float64),
+        )
+
+    def _align_nonbonded_interaction_group(
+        self,
+        src_nb_ixn_group: BoundPotential[NonbondedInteractionGroup],
+        dst_nb_ixn_group: BoundPotential[NonbondedInteractionGroup],
+        combined_atom_map_mixin,
+    ):
+        assert src_nb_ixn_group.potential.beta == dst_nb_ixn_group.potential.beta
+        assert src_nb_ixn_group.potential.cutoff == dst_nb_ixn_group.potential.cutoff
+
+        row_idxs, row_src_params, row_dst_params = self._process_ixn_group_idxs_and_params(
+            src_nb_ixn_group.potential.row_atom_idxs,
+            src_nb_ixn_group.params,
+            dst_nb_ixn_group.potential.row_atom_idxs,
+            dst_nb_ixn_group.params,
+            combined_atom_map_mixin.a_to_c,
+            combined_atom_map_mixin.b_to_c,
+            src_nb_ixn_group.potential.cutoff,
+        )
+
+        col_idxs, col_src_params, col_dst_params = self._process_ixn_group_idxs_and_params(
+            src_nb_ixn_group.potential.col_atom_idxs,
+            src_nb_ixn_group.params,
+            dst_nb_ixn_group.potential.col_atom_idxs,
+            dst_nb_ixn_group.params,
+            combined_atom_map_mixin.a_to_c,
+            combined_atom_map_mixin.b_to_c,
+            src_nb_ixn_group.potential.cutoff,
+        )
+
+        assert len(set(col_idxs).intersection(set(row_idxs))) == 0
+
+        # (ytz): does this have to be jnp?
```

Comment:
To answer this question:

`test_single_topology_combined.py::test_nonbonded_host_params_independent_of_lambda()` fails otherwise.

Reply:
comment inlined.